[fix] quantized_tensor tests

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25429 [fix] quantized_tensor tests**

Summary:
Previously we are using empty to generate test tensors, this PR changes the test tensors to use
randint so that we can test things properly
Also added a set_sizes_and_strides and removed .contiguous() in int_repr function to preserve the
original size and strides

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

Differential Revision: [D17559660](https://our.internmc.facebook.com/intern/diff/D17559660)

jerryzh168 committed:
[fix] quantized_tensor tests

Summary:
Previously we are using empty to generate test tensors, this PR changes the test tensors to use
randint so that we can test things properly
Also added a set_sizes_and_strides and removed .contiguous() in int_repr function to preserve the
original size and strides

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[fix] quantized_tensor tests"


Summary:
Previously we are using empty to generate test tensors, this PR changes the test tensors to use
randint so that we can test things properly
Also added a set_sizes_and_strides and removed .contiguous() in int_repr function to preserve the
original size and strides

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

dzhulgakov commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -100,17 +100,17 @@ Quantizer* quantizer(const Tensor& self) {
+        reinterpret_cast<underlying_t*>(self.data_ptr<scalar_t>());
     underlying_t* dst_data = dst.data_ptr<underlying_t>();
     if (self.numel() > 0) {
       memcpy(dst_data, self_data, self.nbytes());
```
wait a sec, you can't just memcpy if the original tensor is not contiguous. You'd need to use one of the fancy copy functions (probably pull out from cpu implementation of copy_)

jerryzh168 replied:
looks like fancy copy uses TensorIterator. I actually started with TensorIterator, but had some problem with type promotion. Problem is looks like TensorIterator will always compute common type of input and output? https://bddppq.github.io/codebrowser/pytorch/pytorch/aten/src/ATen/native/TensorIterator.cpp.html#125 but what I need is a function from QUInt8 -> UInt8.

dzhulgakov replied:
Hm, not sure. @VitalyFedyunin @li-roy might have a better idea

VitalyFedyunin replied:
Indeed you can't memcpy if your tensor is not contiguous. And indeed TensorIterator currently have problems with in/out types which is @ifedan aiming to solve soon-ish.

dzhulgakov replied:
ok, let's keep it contiguous+memcpy for now then?

ifedan replied:
I'm currently working on PR for TensorIterator, that will allow to compute and propagate common type only from inputs, so output will be able to have any type. Will it solve your problem?

jerryzh168 replied:
@ifedan I think so, please let me know when you land the PR

jerryzh168 replied:
@dzhulgakov we need this change to test view/reshape for quantized tensor, because we need to preserve the memory format when we call int_repr in order to know the memory format are changed properly

ifedan replied:
There is PR for type promotion: https://github.com/pytorch/pytorch/pull/22273. I will wait until it will be landing to continue working on comparison ops.

VitalyFedyunin commented:
#22273 landed, you can revisit this PR.

jerryzh168 committed:
Update on "[fix] quantized_tensor tests"


Summary:
Previously we are using empty to generate test tensors, this PR changes the test tensors to use
randint so that we can test things properly
Also added a set_sizes_and_strides and removed .contiguous() in int_repr function to preserve the
original size and strides

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@VitalyFedyunin looks like we are still trying to promote quantized types to the result type?
```
RuntimeError: promoteTypes with quantized numbers is not handled yet; figure out what the correct rules should be, offending types: Byte QUInt8
```

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -102,17 +104,14 @@ IntArrayRef q_per_channel_axis_quant(const Tensor& self) {
-      memcpy(dst_data, self_data, self.nbytes());
-    }
+    auto iter = TensorIterator::unary_op(dst, self);
+    cpu_kernel(iter, [&](scalar_t value) -> underlying_t {
```
@VitalyFedyunin what I want to do is to get the internal representation of a quantized value,
e.g. `scalar_t` is `qint8` and `underlying_t` is `int8`, looks like the `TensorIterator` always tries to compute the common type of the input and output type first, so the `promoteType` function is called, and we don't have rules to promote a quantized type into a non-quantized type. so why is type promotion needed here?

jerryzh168 replied:
https://bddppq.github.io/codebrowser/pytorch/pytorch/aten/src/ATen/native/TensorIterator.cpp.html#_ZN2atL19compute_result_typeEN3c108ArrayRefINS_11OperandInfoEEERKT_

jerryzh168 committed:
Update on "[fix] quantized_tensor tests"


Summary:
Previously we are using empty to generate test tensors, this PR changes the test tensors to use
randint so that we can test things properly
Also added a set_sizes_and_strides and removed .contiguous() in int_repr function to preserve the
original size and strides

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fix] quantized_tensor tests"


Summary:
Previously we are using empty to generate test tensors, this PR changes the test tensors to use
randint so that we can test things properly
Also added a set_sizes_and_strides and removed .contiguous() in int_repr function to preserve the
original size and strides

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

VitalyFedyunin reviewed (approved):
LGTM for `QTensor.cpp` part

jerryzh168 commented on test/test_quantized_tensor.py:
```diff
@@ -298,28 +302,31 @@ def test_qtensor_view(self):
+        a = torch._per_tensor_affine_qtensor(a_int, scale=scale, zero_point=zero_point)
         b = a.transpose(1, 2)  # swaps 2nd and 3rd dimension
-        c = a.reshape(1, 3, 2, 4)  # does not change tensor layout
+        c = a.reshape(1, 3, 2, 4)  # changes tensor layout in memory
```
Actually I think this is the same as a view right? @gchanan

VitalyFedyunin replied:
View take sizes, reshape takes dimension ids

jerryzh168 replied:
I mean the effect should be the same in this case right?

jerryzh168 replied:
doc for reshape says: `Returns a tensor with the same data and number of elements as input, but with the specified shape. When possible, the returned tensor will be a view of input. Otherwise, it will be a copy.`

jerryzh168 committed:
Update on "[fix] quantized_tensor tests"


Summary:
Previously we are using empty to generate test tensors, this PR changes the test tensors to use
randint so that we can test things properly
Also added a set_sizes_and_strides and removed .contiguous() in int_repr function to preserve the
original size and strides

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
